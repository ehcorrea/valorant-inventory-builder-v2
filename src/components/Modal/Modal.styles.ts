import { ImageBackground, View } from 'react-native';
import Animated from 'react-native-reanimated';
import styled, { css } from 'styled-components/native';

import backgroundImage from '@/assets/images/background-modal.png';

export const Backdrop = styled(Animated.View)`
  background-color: #00000080;
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
`;

export const Modal = styled(Animated.View)`
  ${({ theme }) => css`
    background-color: white;
    border-top-left-radius: ${theme.border.radius.large}px;
    border-top-right-radius: ${theme.border.radius.large}px;
    bottom: 0;
    height: 90%;
    position: absolute;
    width: 100%;
    z-index: 2;
  `}
`;

export const ModalBackground = styled(ImageBackground).attrs(({ theme }) => ({
  source: backgroundImage,
  imageStyle: {
    borderTopLeftRadius: theme.border.radius.large - 1,
    borderTopRightRadius: theme.border.radius.large - 1,
  },
}))`
  flex: 1;
`;

export const ModalContainer = styled(View)`
  ${({ theme }) => css`
    padding-horizontal: ${theme.spacings.large}px;
    padding-top: ${theme.spacings.large}px;
  `}
`;
