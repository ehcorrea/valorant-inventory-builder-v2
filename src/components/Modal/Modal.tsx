import {
  useAnimatedStyle,
  useDerivedValue,
  useSharedValue,
  withDelay,
  withTiming,
} from 'react-native-reanimated';

import * as S from './Modal.styles';

export type ModalProps = {
  children: React.ReactNode;
  open: boolean;
};

export function Modal({ open, children }: ModalProps) {
  const height = useSharedValue(0);
  const progress = useDerivedValue(() =>
    withTiming(Number(!open), { duration: 500 })
  );

  const animatedModal = useAnimatedStyle(() => ({
    transform: [{ translateY: progress.value * 2 * height.value }],
  }));

  const animatedBackdrop = useAnimatedStyle(() => ({
    opacity: 1 - progress.value,
    zIndex: Number(open) || withDelay(500, withTiming(-1, { duration: 0 })),
  }));

  return (
    <>
      {open && (
        <>
          <S.Backdrop style={animatedBackdrop} />
          <S.Modal
            style={animatedModal}
            onLayout={(e) => {
              height.value = e.nativeEvent.layout.height;
            }}
          >
            <S.ModalBackground>
              <S.ModalContainer>{children}</S.ModalContainer>
            </S.ModalBackground>
          </S.Modal>
        </>
      )}
    </>
  );
}
