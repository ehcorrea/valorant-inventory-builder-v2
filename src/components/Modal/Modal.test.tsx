import { View } from 'react-native';

import { render } from '@/test/utils';

import { Modal } from './Modal';

const setup = (open: boolean) =>
  render(
    <Modal open={open}>
      <View testID="child-view" />
    </Modal>
  );

describe('given that the component is rendered', () => {
  test('when open is false, dont show children', async () => {
    const container = setup(false);
    expect(container.queryByTestId('child-view')).toBeFalsy();
    expect(container).toMatchSnapshot();
  });

  test('when open is true, show children ', async () => {
    const container = setup(true);
    expect(container.getByTestId('child-view')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
