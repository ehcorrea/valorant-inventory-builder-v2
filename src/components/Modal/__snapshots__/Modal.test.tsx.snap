// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given that the component is rendered when open is false, dont show children 1`] = `null`;

exports[`given that the component is rendered when open is true, show children  1`] = `
[
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "opacity": 1,
          "zIndex": 1,
        },
      }
    }
    style={
      {
        "backgroundColor": "#00000080",
        "bottom": 0,
        "left": 0,
        "opacity": 1,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 1,
      }
    }
  />,
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "transform": [
            {
              "translateY": 0,
            },
          ],
        },
      }
    }
    onLayout={[Function]}
    style={
      {
        "backgroundColor": "white",
        "borderTopLeftRadius": 20,
        "borderTopRightRadius": 20,
        "bottom": 0,
        "height": "90%",
        "position": "absolute",
        "transform": [
          {
            "translateY": 0,
          },
        ],
        "width": "100%",
        "zIndex": 2,
      }
    }
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        }
      }
    >
      <Image
        source={1}
        style={
          [
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "height": undefined,
              "width": undefined,
            },
            {
              "borderTopLeftRadius": 19,
              "borderTopRightRadius": 19,
            },
          ]
        }
      />
      <View
        style={
          {
            "paddingHorizontal": 20,
            "paddingTop": 20,
          }
        }
      >
        <View
          testID="child-view"
        />
      </View>
    </View>
  </View>,
]
`;
